desc/compute/_bootstrap.py

```diff
@@ -0,0 +1,399 @@
+"""Compute functions for bootstrap current."""
+
+from scipy.constants import elementary_charge
+from scipy.special import roots_legendre
+
+from ..backend import fori_loop, jnp
+from .data_index import register_compute_fun
+from .utils import compress, expand, surface_averages
+
+
+@register_compute_fun(
+    name="trapped fraction",
+    label="1 - \\frac{3}{4} \\left< B^2 \\right> \\int_0^{1/Bmax} "
+    "\\frac{\\lambda\\; d\\lambda}{\\left< \\sqrt{1 - \\lambda B} \\right>}",
+    units="~",
+    units_long="None",
+    description="Neoclassical effective trapped particle fraction",
+    dim=1,
+    params=[],
+    transforms={"grid": []},
+    profiles=[],
+    coordinates="r",
+    data=["sqrt(g)", "V_r(r)", "|B|", "<B^2>", "max_tz |B|"],
+)
+def _trapped_fraction(params, transforms, profiles, data, **kwargs):
+    r"""
+    Evaluate the effective trapped particle fraction.
+
+    Compute the effective fraction of trapped particles, which enters
+    several formulae for neoclassical transport. The trapped fraction
+    ``f_t`` has a standard definition in neoclassical theory:
+
+    .. math::
+        f_t = 1 - \frac{3}{4} \left< B^2 \right> \int_0^{1/Bmax}
+            \frac{\lambda\; d\lambda}{\left< \sqrt{1 - \lambda B} \right>}
+
+    where :math:`\left< \ldots \right>` is a flux surface average.
+    """
+    # Get nodes and weights for Gauss-Legendre integration:
+    n_gauss = kwargs.get("n_gauss", 20)
+    base_nodes, base_weights = roots_legendre(n_gauss)
+    # Rescale for integration on [0, 1], not [-1, 1]:
+    lambd = (base_nodes + 1) * 0.5
+    lambda_weights = base_weights * 0.5
+
+    grid = transforms["grid"]
+    Bmax = data["max_tz |B|"]
+    modB_over_Bmax = data["|B|"] / Bmax
+    sqrt_g = jnp.abs(data["sqrt(g)"])
+    denominator = data["V_r(r)"]
+    Bmax_squared = compress(grid, Bmax * Bmax)
+
+    # Sum over the lambda grid points, using fori_loop for efficiency.
+    lambd = jnp.asarray(lambd)
+    lambda_weights = jnp.asarray(lambda_weights)
+
+    def body_fun(jlambda, lambda_integral):
+        flux_surf_avg_term = surface_averages(
+            grid,
+            jnp.sqrt(1 - lambd[jlambda] * modB_over_Bmax),
+            sqrt_g,
+            denominator=denominator,
+        )
+        return lambda_integral + lambda_weights[jlambda] * lambd[jlambda] / (
+            Bmax_squared * compress(grid, flux_surf_avg_term)
+        )
+
+    lambda_integral = fori_loop(0, n_gauss, body_fun, jnp.zeros(grid.num_rho))
+
+    trapped_fraction = 1 - 0.75 * compress(grid, data["<B^2>"]) * lambda_integral
+    data["trapped fraction"] = expand(grid, trapped_fraction)
+    return data
+
+
+def j_dot_B_Redl(
```

Comment:
There are a couple reasons `j_dot_B_Redl()` was separated out from `_compute_J_dot_B_Redl()`. First, there is that 2nd method for computing Bmin/Bmax/trapped_fraction based on filtering the Boozer spectrum, not yet coded up. That 2nd method would re-use all the code in `j_dot_B_Redl()` but not the code in `_compute_J_dot_B_Redl()`. Second, separating the two functions made it easier to do some tests involving prescribed `trapped fraction` data rather than an actual Equilibrium, specifically `test_Redl_second_pass` and `test_Redl_figures_2_3` and `test_Redl_figures_4_5`. (One way around this could be to directly call the compute functions in the tests, as in `tests/test_bootstrap.py:trapped_fraction()`).